### PR TITLE
Basic SOLR search

### PR DESCRIPTION
**Pull request by kenprice**
SolrSearch
- Should have `$apiUrl` passed as a param to constructor of `SolrSearch`
- Should use `$apiUrl` with a `QueryBuilder` to make a GET request to a SOLR server.
- Should use an `HttpClient` to make GET requests.
- Should parse response from SOLR server, and create a `SearchResult` to return to caller of `SolrSearch#query`.

In addition,
- `SolrQueryBuilder` just returns a dummy query.
- `SearchApi\Clients` has a `HttpClient` interface, intending to be used as a wrapper for whatever HTTP client we use to make our GET requests. For now, `CurlHttpClient` implements this interface and uses curl to make requests. This requires PHP's curl extension.
- SolrSearch has a default `$apiUrl`. This is just a placeholder.
- Spec test for SolrSearch modified to reflect new behavior. `HttpClient#get(...)` is mocked to return a dummy response.

### CODE REVIEW

File: src/clients/curl-http-client.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+   * Simple GET request. Given a URL, return a string for the response.
+   */
+  function get( $url ) {
+    // Use curl to make GET request to SOLR server.
+    $ch = $this->curl;
+    curl_setopt( $ch, CURLOPT_URL, $url );
+    curl_setopt( $ch, CURLOPT_RETURNTRANSFER, 1 );
+    $response = curl_exec( $ch );
+
+    return $response;
```
Comment (rraub) [inline src/clients/curl-http-client.php]: In the future we'll want to handle 404's and other http error states.

File: src/providers/solr-search.php [hunk 2]
```diff
@@ -5,34 +5,97 @@
...
+      $this->apiUrl = 'http://127.0.0.1:8983/solr/gios/select'; // TODO: get this url from a config file
+    }
   }
 
+  /**
+   * Create a search result from a JSON string.
+   * returns SearchResult
+   * @param string $responseString
+   */
+  function parse_query_response( $responseString ) {
```
Comment (rraub) [inline src/providers/solr-search.php:63]: You should at least have two tests that just tests `parse_query_response` by itself. One for a good response string and one with a bad (or empty) response string.

File: src/clients/curl-http-client.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace SearchApi\Clients;
+
+/**
+ * Uses curl to make http requests.
+ */
+class CurlHttpClient implements HttpClient {
```
Comment (rraub) [inline src/clients/curl-http-client.php]: Per converstation: Move this to command

File: src/providers/solr-search.php [hunk 4]
```diff
@@ -5,34 +5,97 @@
...
+  private $apiUrl;
+  private $httpClient;
 
   /**
-   * Constructor
+   * Constructor with optional params
    *
-   * @param queryBuilder QueryBuilder Should be able to inject dependency to use any solr-specific impl. of QueryBuilder
+   * @param QueryBuilder $queryBuilder Should be able to inject dependency to use any solr-specific impl. of QueryBuilder
+   * @param HttpClient $httpClient Preferred http client wrapper. Defaults to CurlHttpClient
```
Comment (rraub) [inline src/providers/solr-search.php]: add a comment for $apiUrl

File: test/spec/SearchApi/Providers/SolrSearchSpec.php [hunk 5]
```diff
@@ -32,10 +32,13 @@ function it_should_return_null_when_keywords_empty() {
     $result->shouldBeNull();
   }
 
-  function it_should_return_search_result_when_keywords_not_empty() {
-    $keywords = array( new Models\SearchTerm( 'test' ) );
-    $result = $this->query( $keywords );
+  function it_should_return_search_result_when_keywords_not_empty( QueryBuilder $queryBuilder, HttpClient $httpClient ) {
+    $this->beConstructedWith( $queryBuilder, $httpClient, 'url/' );
+    $queryBuilder->build( 'test', null )->willReturn( 'test_query' );
+    $httpClient->get( 'url/test_query' )->shouldBeCalled()->willReturn( '{"response":{"numFound":1,"start":0,"docs":[{"id":"testid","author":"testauthor","date":"testdate","content":"testcontent"}]}}' );
```
Comment (rraub) [inline test/spec/SearchApi/Providers/SolrSearchSpec.php]: Move `'{"response":{"numFound":1,"start":0,"docs":[{"id":"testid","author":"testauthor","date":"testdate","content":"testcontent"}]}}'` to its own variable just for readability

File: src/providers/solr-search.php (hunk 2 above)
Comment (kenprice) [inline src/providers/solr-search.php:63]: ✓

File: src/clients/curl-http-client.php (hunk 3 above)
Comment (kenprice) [inline src/clients/curl-http-client.php]: ✓

File: test/spec/SearchApi/Providers/SolrSearchSpec.php (hunk 5 above)
Comment (kenprice) [inline test/spec/SearchApi/Providers/SolrSearchSpec.php]: ✓

File: src/providers/solr-search.php (hunk 4 above)
Comment (kenprice) [inline src/providers/solr-search.php]: ✓

**Pull request merged**